File: plugins/tokenmonitor/server/js/test/provision_providers.test.js
// Regression guard for the 3->2 re-configure provider bug (parity with the Go
// internal/mcp/provision_providers_test.go and py test_provision_providers.py).
//
// When a provision names ANY provider, the broker must forward the WHOLE triple
// so an unchecked provider reaches the device as an explicit `false`. Forwarding
// only the named providers left the device's NVS for the omitted provider
// untouched (the firmware only overwrites keys present in the payload), so a
// device dropped from 3 to 2 providers kept the third enabled.

import { test } from "node:test";
import assert from "node:assert/strict";
import { createServer } from "node:http";

import { provisionTool } from "../src/mcp/server.js";

// Run provisionTool against a throwaway /provision endpoint and resolve with
// the JSON body the broker POSTed to the device.
async function capture(args) {
  const srv = createServer((req, res) => {
    let buf = "";
    req.on("data", (c) => { buf += c; });
    req.on("end", () => {
      srv._body = JSON.parse(buf || "{}");
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end('{"ok":true,"next":"rebooting"}');
    });
  });
  await new Promise((r) => srv.listen(0, "127.0.0.1", r));
  const port = srv.address().port;

  const full = {
    device_id: "ab12cd34",
    provision_url: `http://127.0.0.1:${port}/provision`,
    pairing_code: "071718",
    // Explicit psk_hex keeps provisionTool off the registry-reuse path so a
    // null registry is fine — we only care about the wire body here.
    broker_url: "http://10.0.0.5:8787",
    psk_hex: "0".repeat(64),
    ...args,
  };
  const deps = { cfg: {}, state: {}, logs: null, registry: null, version: "test" };
  await provisionTool(deps, full);
  await new Promise((r) => srv.close(r));
  return srv._body;
}

test("provision drops an unchecked provider (3->2)", async () => {
  // provider_antigravity omitted (user unchecked it) -> must arrive as false.
  const body = await capture({ provider_claude: true, provider_codex: true });
  assert.deepEqual(body.providers, { claude: true, codex: true, gemini: false });
});

test("provision honours the legacy gemini alias and disables the rest", async () => {
  const body = await capture({ provider_gemini: true });
  assert.deepEqual(body.providers, { claude: false, codex: false, gemini: true });
});

test("provision prefers provider_antigravity over the deprecated gemini alias", async () => {
  // Both present: the new arg wins, provider_gemini is ignored.
  const body = await capture({ provider_antigravity: false, provider_gemini: true });
  assert.deepEqual(body.providers, { claude: false, codex: false, gemini: false });
});

test("provision with no provider keys omits the providers field", async () => {
  const body = await capture({ city: "Madrid" });
  assert.equal("providers" in body, false);
});

File: plugins/tokenmonitor/server/go/internal/mcp/provision_providers_test.go
package mcp

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/mark3labs/mcp-go/mcp"
)

// capturedProvision runs handleProvision against a mock device /provision
// endpoint and returns the JSON body the broker POSTed to the device.
func capturedProvision(t *testing.T, args map[string]any) map[string]any {
	t.Helper()

	var body map[string]any
	dev := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		if err := json.Unmarshal(raw, &body); err != nil {
			t.Errorf("device got non-JSON body: %v", err)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"ok":true,"next":"rebooting"}`))
	}))
	defer dev.Close()

	full := map[string]any{
		"device_id":     "ab12cd34",
		"provision_url": dev.URL + "/provision",
		"pairing_code":  "071718",
		// Explicit psk_hex keeps the handler off the registry-reuse path so
		// a nil Registry is fine — we only care about the wire body here.
		"broker_url": "http://10.0.0.5:8787",
		"psk_hex":    "0000000000000000000000000000000000000000000000000000000000000000",
	}
	for k, v := range args {
		full[k] = v
	}

	var req mcp.CallToolRequest
	req.Params.Arguments = full
	if _, err := handleProvision(Deps{})(context.Background(), req); err != nil {
		t.Fatalf("handleProvision: %v", err)
	}
	return body
}

// TestProvision_DropsUncheckedProvider is the regression guard for the
// 3→2 re-configure bug: when a provision names ANY provider, the broker must
// forward the WHOLE triple so an unchecked provider reaches the device as an
// explicit false. Forwarding only the named ones left the device's NVS for the
// omitted provider untouched (the firmware only overwrites keys present in the
// payload), so a device dropped from 3 to 2 providers kept the third enabled.
func TestProvision_DropsUncheckedProvider(t *testing.T) {
	body := capturedProvision(t, map[string]any{
		"provider_claude": true,
		"provider_codex":  true,
		// provider_antigravity intentionally omitted (user unchecked it).
	})

	provs, ok := body["providers"].(map[string]any)
	if !ok {
		t.Fatalf("providers missing or wrong type in POST body: %#v", body["providers"])
	}
	want := map[string]bool{"claude": true, "codex": true, "gemini": false}
	for k, w := range want {
		got, present := provs[k].(bool)
		if !present {
			t.Errorf("providers[%q] absent from wire body — the device would keep its old value", k)
			continue
		}
		if got != w {
			t.Errorf("providers[%q] = %v, want %v", k, got, w)
		}
	}
}

// TestProvision_AntigravityAliasAndAbsent covers the deprecated alias path and
// confirms unnamed providers still default to disabled.
func TestProvision_AntigravityAliasAndAbsent(t *testing.T) {
	body := capturedProvision(t, map[string]any{
		"provider_gemini": true, // legacy alias for antigravity
		// claude and codex omitted → must arrive as explicit false.
	})
	provs, _ := body["providers"].(map[string]any)
	want := map[string]bool{"claude": false, "codex": false, "gemini": true}
	for k, w := range want {
		if got, _ := provs[k].(bool); got != w {
			t.Errorf("providers[%q] = %v, want %v", k, got, w)
		}
	}
}

// TestProvision_AntigravityWinsOverGeminiAlias pins the precedence when a
// caller sends BOTH the new arg and the deprecated alias: provider_antigravity
// wins, provider_gemini is ignored. All three runtimes must agree.
func TestProvision_AntigravityWinsOverGeminiAlias(t *testing.T) {
	body := capturedProvision(t, map[string]any{
		"provider_antigravity": false,
		"provider_gemini":      true, // must be ignored in favour of the new arg
	})
	provs, _ := body["providers"].(map[string]any)
	if got, _ := provs["gemini"].(bool); got != false {
		t.Errorf("providers[gemini] = %v, want false (provider_antigravity must win)", got)
	}
}

// TestProvision_NoProviderKeysOmitsField confirms a partial provision that
// names no provider at all leaves the providers field out entirely, so a
// city-only re-provision does not reset the device's provider selection.
func TestProvision_NoProviderKeysOmitsField(t *testing.T) {
	body := capturedProvision(t, map[string]any{
		"city": "Madrid",
	})
	if _, present := body["providers"]; present {
		t.Errorf("providers should be absent when no provider key is sent, got %#v", body["providers"])
	}
}
